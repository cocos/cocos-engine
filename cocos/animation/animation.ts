/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import './embedded-player/embedded-player';
import './embedded-player/embedded-animation-clip-player';
import './embedded-player/embedded-particle-system-player';
import './tracks/array-track';

export * from './target-path';
export * from './value-proxy';
export { UniformProxyFactory } from './value-proxy-factories/uniform';
export { MorphWeightValueProxy, MorphWeightsValueProxy, MorphWeightsAllValueProxy } from './value-proxy-factories/morph-weights';
export * from './cubic-spline-value';
export { Track, TrackPath } from './tracks/track';
export { RealTrack } from './tracks/real-track';
export { VectorTrack } from './tracks/vector-track';
export { QuatTrack } from './tracks/quat-track';
export { ColorTrack } from './tracks/color-track';
export { SizeTrack } from './tracks/size-track';
export { ObjectTrack } from './tracks/object-track';
export * from './marionette/runtime-exports';
