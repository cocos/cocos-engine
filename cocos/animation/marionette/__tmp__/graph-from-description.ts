/*
 Copyright (c) 2022-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { Vec2 } from '../../../core';
import { AnimationGraph, State, StateMachine, AnimationTransition } from '../animation-graph';
import { Condition, BinaryCondition, TriggerCondition, UnaryCondition } from '../condition';
import { ClipMotion } from '../clip-motion';
import {
    GraphDescription,
    MotionDescription,
    TransitionDescriptionBase,
    StateMachineDescription,
    ParametricDescription,
    AnimationTransitionDescription,
} from './graph-description';
import { AnimationBlend1D } from '../animation-blend-1d';
import { AnimationBlend2D } from '../animation-blend-2d';

import { Motion } from '../motion';
import { Bindable, VariableType } from '../parametric';
import { Value } from '../variable';
import { MotionState } from '../motion-state';

export function createGraphFromDescription (graphDescription: GraphDescription) {
    const graph = new AnimationGraph();

    if (graphDescription.vars) {
        for (const varDesc of graphDescription.vars) {
            switch (getVariableTypeFromValue(varDesc.value)) {
                case VariableType.FLOAT:
                    graph.addFloat(varDesc.name, varDesc.value as number);
                    break;
                case VariableType.BOOLEAN:
                    graph.addBoolean(varDesc.name, varDesc.value as boolean);
                    break;
            }
        }
    }

    for (const layerDesc of graphDescription.layers) {
        const layer = graph.addLayer();
        createSubgraph(layer.stateMachine, layerDesc.graph);
    }
    return graph;
}

function createSubgraph (subgraph: StateMachine, subgraphDesc: StateMachineDescription) {
    const nodes = subgraphDesc.nodes?.map((nodeDesc) => {
        let node: State;
        if (nodeDesc.type === 'animation') {
            const animationState = subgraph.addMotion();
            if (nodeDesc.motion) {
                animationState.motion = createMotion(nodeDesc.motion);
            }
            node = animationState;
        } else {
            const subSubgraph = subgraph.addSubStateMachine();
            createSubgraph(subgraph, nodeDesc);
            node = subSubgraph;
        }
        if (nodeDesc.name) {
            node.name = nodeDesc.name;
        }
        return node;
    }) ?? [];

    if (subgraphDesc.entryTransitions) {
        for (const transitionDesc of subgraphDesc.entryTransitions) {
            createTransition(subgraph, subgraph.entryState, nodes[transitionDesc.to], transitionDesc);
        }
    }
    if (subgraphDesc.exitTransitions) {
        for (const transitionDesc of subgraphDesc.exitTransitions) {
            createAnimationTransition(subgraph, nodes[transitionDesc.from] as MotionState, subgraph.exitState, transitionDesc);
        }
    }
    if (subgraphDesc.anyTransitions) {
        for (const transitionDesc of subgraphDesc.anyTransitions) {
            createTransition(subgraph, subgraph.entryState, nodes[transitionDesc.to], transitionDesc);
        }
    }
    if (subgraphDesc.transitions) {
        for (const transitionDesc of subgraphDesc.transitions) {
            createAnimationTransition(subgraph, nodes[transitionDesc.from] as MotionState, nodes[transitionDesc.to], transitionDesc);
        }
    }
}

function createTransition (graph: StateMachine, from: State, to: State, transitionDesc: TransitionDescriptionBase) {
    let condition: Condition | undefined;
    const conditions = transitionDesc.conditions?.map((conditionDesc) => {
        switch (conditionDesc.type) {
        default:
            throw new Error(`Unknown condition type.`);
        case 'unary': {
            const condition = new UnaryCondition();
            condition.operator = UnaryCondition.Operator[conditionDesc.type];
            createParametric(conditionDesc.operand, condition.operand);
            return condition;
        }
        case 'binary': {
            const condition = new BinaryCondition();
            condition.operator = BinaryCondition.Operator[conditionDesc.type];
            createParametric(conditionDesc.lhs, condition.lhs);
            createParametric(conditionDesc.rhs, condition.rhs);
            return condition;
        }
        case 'trigger': {
            const condition = new TriggerCondition();
            return condition;
        }
        }
    });
    const transition = graph.connect(from, to, conditions);
    return transition;
}

function createAnimationTransition (graph: StateMachine, from: MotionState, to: State, descriptor: AnimationTransitionDescription) {
    const transition = createTransition(graph, from, to, descriptor) as unknown as AnimationTransition;

    const {
        duration,
        exitCondition,
    } = descriptor;

    transition.duration = duration ?? 0.0;

    transition.exitConditionEnabled = false;
    if (typeof exitCondition !== 'undefined') {
        transition.exitConditionEnabled = true;
        transition.exitCondition = exitCondition;
    }
    return transition;
}

function createMotion (motionDesc: MotionDescription): Motion {
    if (motionDesc.type === 'clip') {
        const motion = new ClipMotion();
        return motion;
    } else if (motionDesc.blender.type === '1d') {
        const motion = new AnimationBlend1D();
        const thresholds = motionDesc.blender.thresholds;
        motion.items = motionDesc.children.map((childMotionDesc, iMotion) => {
            const item = new AnimationBlend1D.Item();
            item.motion = createMotion(childMotionDesc);
            item.threshold = thresholds[iMotion];
            return item;
        });
        createParametric(motionDesc.blender.value, motion.param);
        return motion;
    } else {
        const algorithm = AnimationBlend2D.Algorithm[motionDesc.blender.algorithm];
        const motion = new AnimationBlend2D();
        motion.algorithm = algorithm;
        const thresholds = motionDesc.blender.thresholds;
        motion.items = motionDesc.children.map((childMotionDesc, iMotion) => {
            const item = new AnimationBlend2D.Item();
            item.motion = createMotion(childMotionDesc);
            item.threshold = new Vec2(thresholds[iMotion].x, thresholds[iMotion].y);
            return item;
        });
        createParametric(motionDesc.blender.values[0], motion.paramX);
        createParametric(motionDesc.blender.values[1], motion.paramY);
        return motion;
    }
}

function createParametric<T extends string | number | boolean> (paramDesc: ParametricDescription<T>, bindable: Bindable<T>) {
    if (typeof paramDesc === 'object') {
        bindable.variable = paramDesc.name;
        bindable.value = paramDesc.value;
    } else {
        bindable.value = paramDesc;
    }
}

function getVariableTypeFromValue (value: Value) {
    switch (true) {
    case typeof value === 'boolean': return VariableType.BOOLEAN;
    case typeof value === 'number': return VariableType.FLOAT;
    default: throw new Error(`Unknown variable type.`);
    }
}
