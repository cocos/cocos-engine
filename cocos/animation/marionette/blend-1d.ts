export function blend1D (weights: number[], thresholds: readonly number[], value: number) {
    weights.fill(0.0);
    if (thresholds.length === 0) {
        // Do nothing
    } else if (value <= thresholds[0]) {
        weights[0] = 1;
    } else if (value >= thresholds[thresholds.length - 1]) {
        weights[weights.length - 1] = 1;
    } else {
        let iUpper = 0;
        for (let iThresholds = 1; iThresholds < thresholds.length; ++iThresholds) {
            if (thresholds[iThresholds] > value) {
                iUpper = iThresholds;
                break;
            }
        }
        const lower = thresholds[iUpper - 1];
        const upper = thresholds[iUpper];
        const dVal = upper - lower;
        weights[iUpper - 1] = (upper - value) / dVal;
        weights[iUpper] = (value - lower) / dVal;
    }
}
