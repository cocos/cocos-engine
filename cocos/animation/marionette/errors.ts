export class InvalidTransitionError extends Error {
    constructor (type: 'to-entry' | 'to-any' | 'from-exit') {
        super(`${type} transition is invalid`);
        this.name = 'TransitionRejectError';
    }
}

export class VariableNotDefinedError extends Error {
    constructor (name: string) {
        super(`Graph variable ${name} is not defined`);
    }
}

export class VariableTypeMismatchedError extends Error {
    constructor (name: string, expected: string, received?: string) {
        super(`Expect graph variable ${name} to have type '${expected}' instead of received '${received ?? typeof received}'`);
    }
}
