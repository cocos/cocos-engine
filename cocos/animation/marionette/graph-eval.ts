/*
 Copyright (c) 2022-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { DEBUG } from 'internal:constants';
import { AnimationGraph } from './animation-graph';
import type { Node } from '../../scene-graph/node';
import { Value, VarInstance, TriggerResetMode, createInstanceTag, VarInstanceTrigger } from './variable';
import { VariableType } from './parametric';
import { assertIsTrue } from '../../core';
import { MAX_ANIMATION_LAYER } from '../../3d/skeletal-animation/limits';
import type { AnimationController } from './animation-controller';
import {
    AnimationGraphBindingContext, AnimationGraphEvaluationContext,
    AnimationGraphPoseLayoutMaintainer, defaultTransformsTag, LayoutChangeFlag, AuxiliaryCurveRegistry,
    AnimationGraphUpdateContext, AnimationGraphUpdateContextGenerator,
    AnimationGraphSettleContext,
    DeferredPoseStashAllocator,
} from './animation-graph-context';
import { PoseTransformSpaceRequirement } from './pose-graph/pose-node';
import { DefaultTopLevelPoseNode } from './pose-graph/default-top-level-pose-node';
import {
    ClipStatus,
    MotionStateStatus,
    TransitionStatus,
} from './state-machine/state-machine-eval';
import { ReadonlyClipOverrideMap } from './clip-overriding';

export class AnimationGraphEval {
    private _currentTransitionCache: TransitionStatus = {
        duration: 0.0,
        time: 0.0,
    };

    constructor (
        graph: AnimationGraph,
        root: Node,
        controller: AnimationController,
        clipOverrides: ReadonlyClipOverrideMap | null,
    ) {
        if (DEBUG) {
            if (graph.layers.length >= MAX_ANIMATION_LAYER) {
                throw new Error(
                    `Max layer count exceeds. `
                    + `Allowed: ${MAX_ANIMATION_LAYER}, actual: ${graph.layers.length}`,
                );
            }
        }

        for (const [name, variable] of graph.variables) {
            const varInstance = variable[createInstanceTag]();
            this._varInstances[name] = varInstance;
            if (varInstance instanceof VarInstanceTrigger) {
                if (varInstance.resetMode === TriggerResetMode.NEXT_FRAME_OR_AFTER_CONSUMED) {
                    this._hasAutoTrigger = true;
                }
            }
        }

        const poseLayoutMaintainer = new AnimationGraphPoseLayoutMaintainer(root, this._auxiliaryCurveRegistry);
        this._poseLayoutMaintainer = poseLayoutMaintainer;

        const bindingContext = new AnimationGraphBindingContext(root, poseLayoutMaintainer, this._varInstances, controller);
        bindingContext._setClipOverrides(clipOverrides ?? undefined);
        this._bindingContext = bindingContext;

        const settleContext = new AnimationGraphSettleContext(poseLayoutMaintainer);
        this._settleContext = settleContext;

        poseLayoutMaintainer.startBind();

        const poseStashAllocator = new DeferredPoseStashAllocator();
        this._poseStashAllocator = poseStashAllocator;

        this._rootPoseNode = new DefaultTopLevelPoseNode(
            graph,
            bindingContext,
            poseStashAllocator,
        );

        this._root = root;
        this._initializeContexts();
    }

    public destroy (): void {
        this._evaluationContext.destroy();
    }

    public _destroyAfterException_debugging (): void {
        const stackSize = this._evaluationContext._stackSize_debugging;
        if (stackSize !== 0) { // Should only caused by exception.
            for (let i = 0; i < stackSize; ++i) {
                this._evaluationContext.popPose();
            }
        }

        this._evaluationContext.destroy();
    }

    public get layerCount (): number {
        return this._rootPoseNode.layerCount;
    }

    public update (deltaTime: number): void {
        const {
            _evaluationContext: evaluationContext,
            _poseLayoutMaintainer: poseLayoutMaintainer,
            _rootUpdateContextGenerator: rootUpdateContextGenerator,
            _rootPoseNode: rootPoseNode,
        } = this;

        const updateContext = rootUpdateContextGenerator.generate(
            deltaTime,
            1.0,
        );

        rootPoseNode.update(updateContext);

        const finalPose = rootPoseNode.evaluate(evaluationContext, PoseTransformSpaceRequirement.LOCAL);

        if (this._hasAutoTrigger) {
            const { _varInstances: varInstances } = this;
            for (const varName in varInstances) {
                const varInstance = varInstances[varName];
                if (varInstance instanceof VarInstanceTrigger
                    && varInstance.resetMode === TriggerResetMode.NEXT_FRAME_OR_AFTER_CONSUMED) {
                    varInstance.value = false;
                }
            }
        }

        poseLayoutMaintainer.apply(finalPose);
        evaluationContext.popPose();

        if (DEBUG) {
            assertIsTrue(evaluationContext.allocatedPoseCount === 0, `Pose leaked.`);
            assertIsTrue(this._poseStashAllocator.allocatedPoseCount === 0, `Pose leaked.`);
        }
    }

    public getVariables (): Iterable<Readonly<[string, Readonly<{ type: VariableType }>]>> {
        return Object.entries(this._varInstances);
    }

    public getCurrentStateStatus (layer: number): Readonly<MotionStateStatus> | null {
        return this._rootPoseNode.getLayerTopLevelStateMachineEvaluation(layer).getCurrentStateStatus();
    }

    public getCurrentClipStatuses (layer: number): Iterable<Readonly<ClipStatus>> {
        return this._rootPoseNode.getLayerTopLevelStateMachineEvaluation(layer).getCurrentClipStatuses();
    }

    public getCurrentTransition (layer: number): Readonly<TransitionStatus> | null {
        const {
            _currentTransitionCache: currentTransition,
        } = this;
        const isInTransition = this._rootPoseNode.getLayerTopLevelStateMachineEvaluation(layer).getCurrentTransition(currentTransition);
        return isInTransition ? currentTransition : null;
    }

    public getNextStateStatus (layer: number): Readonly<MotionStateStatus> | null {
        return this._rootPoseNode.getLayerTopLevelStateMachineEvaluation(layer).getNextStateStatus();
    }

    public getNextClipStatuses (layer: number): Iterable<Readonly<ClipStatus>> {
        return this._rootPoseNode.getLayerTopLevelStateMachineEvaluation(layer).getNextClipStatuses();
    }

    public getValue (name: string): Value | undefined {
        const varInstance = this._varInstances[name];
        if (!varInstance) {
            return undefined;
        } else {
            return varInstance.value;
        }
    }

    public setValue (name: string, value: Value): void {
        const varInstance = this._varInstances[name];
        if (!varInstance) {
            return;
        }
        varInstance.value = value;
    }

    public getLayerWeight (layerIndex: number): number {
        return this._rootPoseNode.getLayerWeight(layerIndex);
    }

    public setLayerWeight (layerIndex: number, weight: number): void {
        this._rootPoseNode.setLayerWeight(layerIndex, weight);
    }

    public overrideClips (overrides: ReadonlyClipOverrideMap): void {
        const {
            _poseLayoutMaintainer: poseLayoutMaintainer,
        } = this;

        poseLayoutMaintainer.startBind();

        this._bindingContext._setClipOverrides(overrides);
        this._rootPoseNode.overrideClips(this._bindingContext);

        this._updateAfterPossiblePoseLayoutChange();
    }

    public getAuxiliaryCurveValue (curveName: string): number {
        return this._auxiliaryCurveRegistry.get(curveName);
    }

    private _rootPoseNode: DefaultTopLevelPoseNode;
    private _varInstances: Record<string, VarInstance> = {};
    private _hasAutoTrigger = false;
    private _auxiliaryCurveRegistry = new AuxiliaryCurveRegistry();
    private _poseLayoutMaintainer: AnimationGraphPoseLayoutMaintainer;
    private _bindingContext: AnimationGraphBindingContext;
    private _settleContext: AnimationGraphSettleContext;
    /**
     * Preserved here for clip overriding.
     */
    private declare _root: Node;
    private declare _evaluationContext: AnimationGraphEvaluationContext;
    private declare _poseStashAllocator: DeferredPoseStashAllocator;
    private _rootUpdateContextGenerator = new AnimationGraphUpdateContextGenerator();

    private _initializeContexts (): void {
        const {
            _poseLayoutMaintainer: poseLayoutMaintainer,
        } = this;

        // Ignore in initialization.
        // eslint-disable-next-line no-void
        void poseLayoutMaintainer.endBind();

        this._createOrUpdateTransformFilters();

        const evaluationContext = poseLayoutMaintainer.createEvaluationContext();
        this._evaluationContext = evaluationContext;

        // Capture the default transforms.
        poseLayoutMaintainer.fetchDefaultTransforms(evaluationContext[defaultTransformsTag]);

        poseLayoutMaintainer.resetPoseStashAllocator(this._poseStashAllocator);
    }

    private _updateAfterPossiblePoseLayoutChange (): void {
        const {
            _poseLayoutMaintainer: poseLayoutMaintainer,
        } = this;

        const layoutChangeFlags = poseLayoutMaintainer.endBind();

        // Nothing changed, this should be the commonest case in real world.
        if (layoutChangeFlags === 0) {
            return;
        }

        // No matter count or order changed, we should update the transform filters.
        if ((layoutChangeFlags & LayoutChangeFlag.TRANSFORM_COUNT)
            || (layoutChangeFlags & LayoutChangeFlag.TRANSFORM_ORDER)) {
            this._createOrUpdateTransformFilters();
        }

        // Either transform count or auxiliary curve count changed, we should recreate the eval context.
        let evaluationContextRecreated = false;
        if ((layoutChangeFlags & LayoutChangeFlag.TRANSFORM_COUNT)
        || (layoutChangeFlags & LayoutChangeFlag.AUXILIARY_CURVE_COUNT)) {
            const evaluationContext = poseLayoutMaintainer.createEvaluationContext();
            this._evaluationContext.destroy();
            this._evaluationContext = evaluationContext;
            evaluationContextRecreated = true;
            poseLayoutMaintainer.resetPoseStashAllocator(this._poseStashAllocator);
        }

        // If the eval context was recreated or the layout has changed, we should update the default transforms.
        if (evaluationContextRecreated
            || (layoutChangeFlags & LayoutChangeFlag.TRANSFORM_COUNT)
            || (layoutChangeFlags & LayoutChangeFlag.TRANSFORM_ORDER)) {
            poseLayoutMaintainer.fetchDefaultTransforms(this._evaluationContext[defaultTransformsTag]);
        }
    }

    private _createOrUpdateTransformFilters (): void {
        this._rootPoseNode.settle(this._settleContext);
    }
}

export type { VarInstance } from './variable';
