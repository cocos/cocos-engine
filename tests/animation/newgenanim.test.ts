
import { AnimationClip, Node, Vec2 } from '../../cocos/core';
import { PoseBlend1D, PoseBlend2D, Condition, InvalidTransitionError, VariableNotDefinedError, __getDemoGraphs } from '../../cocos/core/animation/animation';
import { PoseGraph } from '../../cocos/core/animation/newgen-anim/pose-graph';
import { createEval } from '../../cocos/core/animation/newgen-anim/create-eval';
import { VariableTypeMismatchedError } from '../../cocos/core/animation/newgen-anim/errors';
import { PoseGraphEval } from '../../cocos/core/animation/newgen-anim/graph-eval';
import { createGraphFromDescription } from '../../cocos/core/animation/newgen-anim/__tmp__/graph-from-description';
import gAnyTransition from './graphs/any-transition';
import gUnspecifiedCondition from './graphs/unspecified-condition';
import gSuccessiveSatisfaction from './graphs/successive-satisfaction';
import gVariableNotFoundInCondition from './graphs/variable-not-found-in-condition';
import gVariableNotFoundInPoseBlend from './graphs/variable-not-found-in-pose-blend';
import gPoseBlendRequiresNumbers from './graphs/pose-blend-requires-numbers';
import { getPropertyBindingPoints } from '../../cocos/core/animation/newgen-anim/parametric';
import { blend1D } from '../../cocos/core/animation/newgen-anim/blend-1d';

describe('NewGen Anim', () => {
    const demoGraphs = __getDemoGraphs();

    describe('Connecting', () => {
        test('Connecting', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            const n1 = layerGraph.add();
            const n2 = layerGraph.add();
            layerGraph.connect(n1, n2);
            expect([...layerGraph.getOutgoings(n1)].map((t) => t.to)).toContain(n2);
            expect([...layerGraph.getIncomings(n2)].map((t) => t.from)).toContain(n1);
        });

        test('Reconnecting', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            const n1 = layerGraph.add();
            const n2 = layerGraph.add();
            const trans1 = layerGraph.connect(n1, n2);
            const trans2 = layerGraph.connect(n1, n2);
            expect(trans1).not.toBe(trans2);
            expect(layerGraph.getTransition(n1, n2)).toBe(trans2);
        });

        test('Self connecting', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            const n1 = layerGraph.add();
            layerGraph.connect(n1, n1);
            // TODO: what's the expectation?
        });
    });

    describe('Transitions', () => {
        test('Could not transition to entry node', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            expect(() => layerGraph.connect(layerGraph.add(), layerGraph.entryNode)).toThrowError(InvalidTransitionError);
        });

        test('Could not transition to any node', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            expect(() => layerGraph.connect(layerGraph.add(), layerGraph.anyNode)).toThrowError(InvalidTransitionError);
        });

        test('Could not transition from exit node', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            expect(() => layerGraph.connect(layerGraph.existNode, layerGraph.add())).toThrowError(InvalidTransitionError);
        });

        test('Condition not specified', () => {
            const graphEval = new PoseGraphEval(createGraphFromDescription(gUnspecifiedCondition), new Node());
            graphEval.update(0.0);
            expect(graphEval.getCurrentNodeInfo(0).name).toBe('asd');
        });

        test('One transition most per update', () => {
            const graphEval = new PoseGraphEval(createGraphFromDescription(gSuccessiveSatisfaction), new Node());
            graphEval.update(0.0);
            expect(graphEval.getCurrentNodeInfo(0).name).toBe('Node1');
            graphEval.update(0.0);
            expect(graphEval.getCurrentNodeInfo(0).name).toBe('Node2'); 
        });

        test('Any transition', () => {
            const graphEval = new PoseGraphEval(createGraphFromDescription(gAnyTransition), new Node());
            graphEval.update(0.0);
            expect(graphEval.getCurrentNodeInfo(0).name).toBe('Node1');
        });
    });

    describe('Removing nodes', () => {
        test('Could not remove special nodes', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            layerGraph.remove(layerGraph.entryNode);
            layerGraph.remove(layerGraph.existNode);
            layerGraph.remove(layerGraph.anyNode);
            expect([...layerGraph.nodes()]).toEqual(expect.arrayContaining([
                layerGraph.entryNode,
                layerGraph.existNode,
                layerGraph.anyNode,
            ]));
        });

        test('Also erase referred transitions', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            const node1 = layerGraph.add();
            const node2 = layerGraph.add();
            const node3 = layerGraph.add();
            layerGraph.connect(node1, node2);
            layerGraph.connect(node3, node1);

            layerGraph.remove(node2);
            expect([...layerGraph.getOutgoings(node1)]).toHaveLength(0);
            layerGraph.remove(node3);
            expect([...layerGraph.getIncomings(node1)]).toHaveLength(0);
        });
    });

    describe('Exotic nodes', () => {
        test('Removed nodes are dangling', () => {
            const graph = new PoseGraph();
            const layer = graph.addLayer();
            const layerGraph = layer.graph;
            const node = layerGraph.add();
            layerGraph.remove(node);
            expect(() => layerGraph.remove(node)).toThrow();
            expect(() => layerGraph.getIncomings(node)).toThrow();
            expect(() => layerGraph.connect(layerGraph.entryNode, node)).toThrow();
        });

        test('Nodes in different layers are isolated', () => {
            const graph = new PoseGraph();
            const layer1 = graph.addLayer();
            const layerGraph1 = layer1.graph;
            const layer2 = graph.addLayer();
            const layerGraph2 = layer2.graph;
            const node1 = layerGraph1.add();
            const node2 = layerGraph2.add();
            expect(() => layerGraph2.connect(node2, node1)).toThrow();
        });
    });

    describe('Condition', () => {
        const emptyContext: Parameters<Condition[typeof createEval]>[0] = { getParam() {} };
        test('Unary', () => {
            const condition = new Condition();
            condition.operator = Condition.Operator.BE_TRUE;
            condition.lhs = false;
            const conditionEval = condition[createEval](emptyContext);
            expect(conditionEval.eval()).toBe(false);
            conditionEval.setLhs(true);
            expect(conditionEval.eval()).toBe(true);
        });

        test('Binary', () => {
            const condition = new Condition();
            condition.operator = Condition.Operator.GREATER_THAN;
            condition.lhs = 1;
            condition.rhs = 0.5;
            const conditionEval = condition[createEval](emptyContext);
            expect(conditionEval.eval()).toBe(true);
            conditionEval.setRhs(2.0);
            expect(conditionEval.eval()).toBe(false);
        });
    });

    describe('Blender 1D', () => {
        test('Thresholds should have been sorted', () => {
            const blender1D = new PoseBlend1D();
            blender1D.children = [[null, 0.3], [null, 0.2]];
            expect([...blender1D.children].map(([, threshold]) => threshold)).toStrictEqual([0.2, 0.3]);

            blender1D.children = [[null, 0.9], [null, -0.2]];
            expect([...blender1D.children].map(([, threshold]) => threshold)).toStrictEqual([-0.2, 0.9]);
        });

        test('1D Blending', () => {
            const thresholds = [0.1, 0.3] as const;
            const weights = new Array<number>(thresholds.length).fill(0);

            blend1D(weights, thresholds, 0.4);
            expect(weights).toBeCloseToArray([0.0, 1.0]);

            blend1D(weights, thresholds, 0.1);
            expect(weights).toBeCloseToArray([1.0, 0.0]);

            blend1D(weights, thresholds, 0.3);
            expect(weights).toBeCloseToArray([0.0, 1.0]);

            blend1D(weights, thresholds, 0.2);
            expect(weights).toBeCloseToArray([0.5, 0.5]);
        });
    });

    describe('Variable not found error', () => {
        test('Missed in conditions', () => {
            expect(() => new PoseGraphEval(createGraphFromDescription(gVariableNotFoundInCondition), new Node())).toThrowError(VariableNotDefinedError);
        });

        test('Missed in pose blend', () => {
            expect(() => new PoseGraphEval(createGraphFromDescription(gVariableNotFoundInPoseBlend), new Node())).toThrowError(VariableNotDefinedError);
        });
    });

    describe('Variable type mismatch error', () => {
        test('pose blend requires numbers', () => {
            expect(() => new PoseGraphEval(createGraphFromDescription(gPoseBlendRequiresNumbers), new Node())).toThrowError(VariableTypeMismatchedError);
        });
    });

    describe('Property binding', () => {
        test('Bind property', () => {
            const poseBlend2D = new PoseBlend2D();
            const bindingPoints = getPropertyBindingPoints(poseBlend2D);
            expect(Object.keys(bindingPoints)).toEqual(expect.arrayContaining([
                'paramX',
                'paramY',
            ]));
            poseBlend2D.bindProperty('paramX', 'x');
            expect(poseBlend2D.getPropertyBinding('paramX')).toBe('x');
        });

        test('Serialization', () => {
            const poseBlend2D = new PoseBlend2D();
            poseBlend2D.bindProperty('paramX', 'x');
            expect(poseBlend2D.getPropertyBinding('paramX')).toBe('x');
        });
    });
});
