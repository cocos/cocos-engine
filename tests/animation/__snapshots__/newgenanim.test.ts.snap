// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewGen Anim Animation mask Multiple layers _ 1`] = `
Object {
  "1": 1.1688000000000003,
  "1_1": 0.7440000000000001,
  "1_1_1": 1.3688000000000002,
  "1_2": 1.3920000000000001,
}
`;

exports[`NewGen Anim Animation mask Single layer Leaf is masked 1`] = `
Object {
  "1": 0.42000000000000004,
  "1_1": 0.52,
  "1_1_1": 0.30000000000000004,
  "1_2": 0.7200000000000001,
}
`;

exports[`NewGen Anim Animation mask Single layer Middle is masked 1`] = `
Object {
  "1": 0.42000000000000004,
  "1_1": 0.2,
  "1_1_1": 0.62,
  "1_2": 0.7200000000000001,
}
`;

exports[`NewGen Anim Animation mask Single layer Root and leaf are masked 1`] = `
Object {
  "1": 0.1,
  "1_1": 0.52,
  "1_1_1": 0.30000000000000004,
  "1_2": 0.7200000000000001,
}
`;

exports[`NewGen Anim Animation mask Single layer Root is masked 1`] = `
Object {
  "1": 0.1,
  "1_1": 0.52,
  "1_1_1": 0.62,
  "1_2": 0.7200000000000001,
}
`;

exports[`NewGen Anim Transitions Empty->Empty transition: Expected result 1`] = `0.7`;
